lib/proxy/conn.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/trace"
+	"google.golang.org/grpc"
+)
+
+// Stream is a common interface for grpc client and server streams.
+type Stream interface {
+	Context() context.Context
+	Send(*proto.Frame) error
+	Recv() (*proto.Frame, error)
+}
+
+// streamConn wraps a grpc stream with a net.streamConn interface.
+type streamConn struct {
+	stream Stream
+
+	wLock  sync.Mutex
+	rLock  sync.Mutex
+	rBytes []byte
+	rOff   int
+
+	src net.Addr
+	dst net.Addr
+}
+
+// newStreamConn creates a new streamConn.
+func newStreamConn(stream Stream, src net.Addr, dst net.Addr) *streamConn {
+	return &streamConn{
+		stream: stream,
+		src:    src,
+		dst:    dst,
+	}
+}
+
+// Read reads data reveived over the grpc stream.
+func (c *streamConn) Read(b []byte) (n int, err error) {
+	c.rLock.Lock()
```

Comment:
```suggestion
func (c *streamConn) Read(b []byte) (n int, err error) {
	if len(b) == 0 {
		return 0, nil
	}
	c.rLock.Lock()
```

Reply:
I would rather let this through that way it will fail if the stream is closed.

Reply:
A zero read would still be a noop if we had a non-empty buffer ready.